File: backend/src/main/java/org/example/studentmanagement/service/ClassSessionService.java
package org.example.studentmanagement.service;

import org.example.studentmanagement.entity.ClassSession;
import org.example.studentmanagement.entity.Etudiant;
import org.example.studentmanagement.repository.ClassSessionRepository;
import org.example.studentmanagement.repository.EtudiantRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.DayOfWeek;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class ClassSessionService {

    @Autowired
    private ClassSessionRepository classSessionRepository;

    @Autowired
    private EtudiantRepository etudiantRepository;

    public List<ClassSession> getAllClassSessions() {
        return classSessionRepository.findAll();
    }

    public Optional<ClassSession> getClassSessionById(Long id) {
        return classSessionRepository.findById(id);
    }

    public List<ClassSession> getClassSessionsByStudentId(Long studentId) {
        return classSessionRepository.findByStudentId(studentId);
    }

    public List<ClassSession> getClassSessionsByDayOfWeek(DayOfWeek dayOfWeek) {
        return classSessionRepository.findByDayOfWeek(dayOfWeek);
    }

    public List<ClassSession> getClassSessionsByStudentAndDayOfWeek(Long studentId, DayOfWeek dayOfWeek) {
        return classSessionRepository.findByStudentIdAndDayOfWeek(studentId, dayOfWeek);
    }

    public List<ClassSession> getClassSessionsBySubjectId(Long subjectId) {
        return classSessionRepository.findBySubjectId(subjectId);
    }

    public ClassSession createClassSession(ClassSession classSession) {
        // Validate that class sessions are not on weekends
        if (classSession.getDayOfWeek() == DayOfWeek.SATURDAY || classSession.getDayOfWeek() == DayOfWeek.SUNDAY) {
            throw new IllegalArgumentException("Class sessions cannot be scheduled on weekends");
        }
        
        // Validate that the class session doesn't overlap with existing sessions in the same room
        validateClassSessionOverlap(classSession);
        
        // Validate that end time is after start time
        if (classSession.getEndTime().isBefore(classSession.getStartTime()) || 
            classSession.getEndTime().equals(classSession.getStartTime())) {
            throw new IllegalArgumentException("End time must be after start time");
        }
        
        return classSessionRepository.save(classSession);
    }

    public ClassSession updateClassSession(Long id, ClassSession classSessionDetails) {
        Optional<ClassSession> classSessionOptional = classSessionRepository.findById(id);
        if (classSessionOptional.isPresent()) {
            ClassSession classSession = classSessionOptional.get();
            
            // Validate that class sessions are not on weekends
            if (classSessionDetails.getDayOfWeek() == DayOfWeek.SATURDAY || 
                classSessionDetails.getDayOfWeek() == DayOfWeek.SUNDAY) {
                throw new IllegalArgumentException("Class sessions cannot be scheduled on weekends");
            }
            
            // Validate that the class session doesn't overlap with existing sessions in the same room
            validateClassSessionOverlap(classSessionDetails);
            
            // Validate that end time is after start time
            if (classSessionDetails.getEndTime().isBefore(classSessionDetails.getStartTime()) || 
                classSessionDetails.getEndTime().equals(classSessionDetails.getStartTime())) {
                throw new IllegalArgumentException("End time must be after start time");
            }
            
            classSession.setSubject(classSessionDetails.getSubject());
            classSession.setInstructor(classSessionDetails.getInstructor());
            classSession.setDayOfWeek(classSessionDetails.getDayOfWeek());
            classSession.setStartTime(classSessionDetails.getStartTime());
            classSession.setEndTime(classSessionDetails.getEndTime());
            classSession.setRoom(classSessionDetails.getRoom());
            classSession.setStudents(classSessionDetails.getStudents());
            
            return classSessionRepository.save(classSession);
        }
        throw new RuntimeException("Class session not found with id: " + id);
    }

    public void deleteClassSession(Long id) {
        classSessionRepository.deleteById(id);
    }

    public void addStudentToClassSession(Long classSessionId, Long studentId) {
        Optional<ClassSession> classSessionOptional = classSessionRepository.findById(classSessionId);
        Optional<Etudiant> studentOptional = etudiantRepository.findById(studentId);
        
        if (classSessionOptional.isPresent() && studentOptional.isPresent()) {
            ClassSession classSession = classSessionOptional.get();
            Etudiant student = studentOptional.get();
            
            if (!classSession.getStudents().contains(student)) {
                classSession.getStudents().add(student);
                classSessionRepository.save(classSession);
            }
        }
    }

    public void removeStudentFromClassSession(Long classSessionId, Long studentId) {
        Optional<ClassSession> classSessionOptional = classSessionRepository.findById(classSessionId);
        Optional<Etudiant> studentOptional = etudiantRepository.findById(studentId);
        
        if (classSessionOptional.isPresent() && studentOptional.isPresent()) {
            ClassSession classSession = classSessionOptional.get();
            Etudiant student = studentOptional.get();
            
            classSession.getStudents().remove(student);
            classSessionRepository.save(classSession);
        }
    }

    private void validateClassSessionOverlap(ClassSession classSession) {
        List<ClassSession> overlappingSessions = classSessionRepository.findOverlappingClassSessions(
            classSession.getRoom(),
            classSession.getDayOfWeek(),
            classSession.getStartTime().toString(),
            classSession.getEndTime().toString()
        );
        
        // Remove the current class session from the list if it's an update
        overlappingSessions.removeIf(cs -> cs.getId().equals(classSession.getId()));
        
        if (!overlappingSessions.isEmpty()) {
            throw new IllegalArgumentException("Class session overlaps with existing sessions in the same room and time");
        }
    }
} 

File: frontend/src/app/services/class-session.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ClassSession, ClassSessionFormData, DayOfWeek } from '../models/class-session.model';

@Injectable({
  providedIn: 'root'
})
export class ClassSessionService {
  private readonly API_URL = 'http://localhost:8080/api/class-sessions';

  constructor(private http: HttpClient) { }

  // Admin methods
  getAllClassSessions(): Observable<ClassSession[]> {
    return this.http.get<ClassSession[]>(this.API_URL);
  }

  getClassSessionById(id: number): Observable<ClassSession> {
    return this.http.get<ClassSession>(`${this.API_URL}/${id}`);
  }

  createClassSession(classSessionData: ClassSessionFormData): Observable<ClassSession> {
    return this.http.post<ClassSession>(this.API_URL, classSessionData);
  }

  updateClassSession(id: number, classSessionData: ClassSessionFormData): Observable<ClassSession> {
    return this.http.put<ClassSession>(`${this.API_URL}/${id}`, classSessionData);
  }

  deleteClassSession(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  addStudentToClassSession(classSessionId: number, studentId: number): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/${classSessionId}/students/${studentId}`, {});
  }

  removeStudentFromClassSession(classSessionId: number, studentId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${classSessionId}/students/${studentId}`);
  }

  // Student methods
  getClassSessionsByStudentId(studentId: number): Observable<ClassSession[]> {
    return this.http.get<ClassSession[]>(`${this.API_URL}/student/${studentId}`);
  }

  getClassSessionsByStudentAndDayOfWeek(studentId: number, dayOfWeek: DayOfWeek): Observable<ClassSession[]> {
    return this.http.get<ClassSession[]>(`${this.API_URL}/student/${studentId}/day/${dayOfWeek}`);
  }

  // Additional filtering methods
  getClassSessionsByDayOfWeek(dayOfWeek: DayOfWeek): Observable<ClassSession[]> {
    return this.http.get<ClassSession[]>(`${this.API_URL}/day/${dayOfWeek}`);
  }

  getClassSessionsBySubjectId(subjectId: number): Observable<ClassSession[]> {
    return this.http.get<ClassSession[]>(`${this.API_URL}/subject/${subjectId}`);
  }
} 